migrations/entitlements/migration.go

```diff
@@ -0,0 +1,245 @@
+		)
+	case *interpreter.StorageReferenceValue:
+		staticType = interpreter.NewReferenceStaticType(
+			inter,
+			referenceValue.Authorization,
+			interpreter.ConvertSemaToStaticType(inter, referenceValue.BorrowedType),
+		)
+	default:
+		staticType = v.StaticType(inter)
+	}
+
+	// if the static type contains a legacy restricted type, convert it to a new type according to some rules:
+	// &T{I} -> auth(SupportedEntitlements(I)) &T
+	// Capability<&T{I}> -> Capability<auth(SupportedEntitlements(I)) &T>
+	var convertLegacyStaticType func(interpreter.StaticType)
+	convertLegacyStaticType = func(staticType interpreter.StaticType) {
+		switch t := staticType.(type) {
+		case *interpreter.ReferenceStaticType:
+			switch referencedType := t.ReferencedType.(type) {
+			case *interpreter.IntersectionStaticType:
+				if referencedType.LegacyType != nil {
+					t.ReferencedType = referencedType.LegacyType
+					intersectionSemaType := inter.MustConvertStaticToSemaType(referencedType).(*sema.IntersectionType)
+					auth := sema.UnauthorizedAccess
+					supportedEntitlements := intersectionSemaType.SupportedEntitlements()
+					if supportedEntitlements.Len() > 0 {
+						auth = sema.EntitlementSetAccess{
+							SetKind:      sema.Conjunction,
+							Entitlements: supportedEntitlements,
+						}
+					}
+					t.Authorization = interpreter.ConvertSemaAccessToStaticAuthorization(inter, auth)
+				}
+			}
+		case *interpreter.CapabilityStaticType:
+			convertLegacyStaticType(t.BorrowType)
+		case *interpreter.VariableSizedStaticType:
+			convertLegacyStaticType(t.Type)
+		case *interpreter.ConstantSizedStaticType:
+			convertLegacyStaticType(t.Type)
+		case *interpreter.DictionaryStaticType:
+			convertLegacyStaticType(t.KeyType)
+			convertLegacyStaticType(t.ValueType)
+		case *interpreter.OptionalStaticType:
+			convertLegacyStaticType(t.Type)
+		}
+	}
+
+	convertLegacyStaticType(staticType)
+	semaType := inter.MustConvertStaticToSemaType(staticType)
+	entitledType := ConvertToEntitledType(semaType)
+
+	// if the types of the values are equal and the value is not a runtime type, there's nothing to migrate
+	if entitledType.Equal(semaType) && !entitledType.Equal(sema.MetaType) {
+		return nil
+	}
+
+	switch v := v.(type) {
+	case *interpreter.EphemeralReferenceValue:
+		entitledReferenceType := entitledType.(*sema.ReferenceType)
+		staticAuthorization := interpreter.ConvertSemaAccessToStaticAuthorization(inter, entitledReferenceType.Authorization)
+		convertedValue := ConvertValueToEntitlements(inter, v.Value)
+		// if the underlying value did not change, we still want to use the old value in the newly created reference
+		if convertedValue == nil {
+			convertedValue = v.Value
+		}
+		return interpreter.NewEphemeralReferenceValue(
+			inter,
+			staticAuthorization,
+			convertedValue,
+			entitledReferenceType.Type,
+			interpreter.EmptyLocationRange,
+		)
+
+	case *interpreter.StorageReferenceValue:
```

Comment:
Same as above: These two cases, ephemeral reference values and storage reference values, should be unreachable

Reply:
These are here for the same reason as above.

Reply:
Makes sense! Could you please leave a comment in the code?